File: problem_12.cpp
#include <chrono>
#include <iostream>
#include <vector>
#include <cstring>
#include <cmath>
#include <functional>

const long limitNumber = 1000000;
const int targetFactors = 500;

long triangular (long n) 
{
    return (n * (n + 1)) / 2;
}

void allPrimesUsingSieve (long n, std::vector<long>* primes) 
{
    std::vector<unsigned char> sieve;
    sieve.resize(n + 1);
    memset(sieve.data(), 0, (n + 1) * sizeof(unsigned char));
    for (long i = 2; i <= int(sqrt(n)); i++) 
    {
        if (sieve[i] != 0) 
        {
            continue;
        }
        for (long j = i + i; j < n; j += i) 
        {
            sieve[j] = (unsigned short) 1;
        }
    } 
    for (long i = 2; i < n; i++) 
    {
        if (sieve[i] == 0)
        {
            primes->push_back(i);
        }
    }
}

void factorize(int n, std::vector<long>* primes, std::vector<int>* exponents)
{
    for (int i = 0; i < primes->size(); i++)
    {
        long curPrime = primes->at(i);
        if (curPrime > sqrt(n)) 
        {
            break;
        }
        int curExponent = 0;
        while (n % curPrime == 0 && curPrime <= n) 
        {
            curExponent++;
            curPrime *= primes->at(i);
        }
        exponents->at(i) = curExponent;
    }
}

int main() 
{
    auto startTm = std::chrono::high_resolution_clock::now();

    std::vector<long> primes;
    allPrimesUsingSieve(limitNumber, &primes);
    std::vector<int> primeExponents;
    primeExponents.resize(primes.size());
    int numberOfFactors = 0;
    long candidate;
    int i = 0;
    while (numberOfFactors < targetFactors && i < limitNumber) 
    {
        memset(primeExponents.data(), 0, (primes.size()) * sizeof(int));
        candidate = triangular(i);
        factorize(candidate, &primes, &primeExponents);
        i++;
        numberOfFactors = 1;
        for (int j = 0; j < primes.size(); j++) 
        {  
            if (primes.at(j) > sqrt(candidate))
            {
                break;
            }
            numberOfFactors *= primeExponents.at(j) + 1;
        }
    }
    
    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Solution: " << candidate << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_12.rs
use std::time::{Instant};

const LIMIT_NUMBER: usize = 1000000;
const TARGET_FACTORS: usize = 500;

fn triangular(n: usize) -> usize
{
    (n * (n + 1)) / 2
}

fn all_primes_using_sieve (n: usize) -> Vec<usize>
{
    let mut sieve: Vec<u8> = vec![];
    let mut primes: Vec<usize> = vec![];
    sieve.resize(n + 1, 0);
    let limit = f64::sqrt(n as f64) as usize;
    for i in 2..limit 
    {
        if sieve[i] == 0 
        {
            for j in ((i + i)..n).step_by(i)
            {
                sieve[j] = 1;
            }
        }
    }
    for i in 2..n 
    {
        if sieve[i] == 0
        {
            primes.push(i);
        }
    }
    return primes;
}

fn factorize(n : usize, primes: &Vec<usize>) -> Vec<usize> 
{
    let mut exponents: Vec<usize> = vec![0usize; primes.len()];
    for i in 0..primes.len()
    {
        if primes[i] > f64::sqrt(n as f64) as usize 
        {
            break;
        }
        let mut cur: usize = primes[i];
        let mut cur_exponent: usize = 0;
        while n % cur == 0 && cur <= n 
        {
            cur_exponent = cur_exponent + 1;
            cur = cur * primes[i];
        }
        exponents[i] = cur_exponent;
    }
    return exponents;
}


fn main () 
{
    let start = Instant::now();

    let primes: Vec<usize> = all_primes_using_sieve(LIMIT_NUMBER);
    for i in 0..LIMIT_NUMBER
    {
        let candidate: usize = triangular(i);
        let exponents = factorize(candidate, &primes);
        let mut number_of_factors: usize = 1;
        for j in 0..primes.len()
        {  
            if primes[j] > f64::sqrt(candidate as f64) as usize
            {
                break;
            }
            number_of_factors = number_of_factors * (exponents[j] + 1);
        }
        if number_of_factors >= TARGET_FACTORS
        {
            println!("Result: {}", candidate);
            break;
        }
    }
    
    let duration = start.elapsed();
    println!("Timing: {}", duration.as_nanos());
}


